### PR TITLE
support custom issuer with external account binding and domain restrictions

### PR DESCRIPTION
**Pull request by MartinWeindel**
**How to categorize this PR?**

/area control-plane
/kind enhancement
/priority normal

**What this PR does / why we need it**:
This PR provides adjustments for the v0.5.0 release of cert management:
- updated CRDs for issuers, certificates and new CRD for certificaterevocations
- support for custom issuer with external account binding
- support for domain restrictions in custom issuer spec
- support for providing private key for custom issuer via shoot resource reference

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature user
support for custom issuer with external account binding
```

```feature user
support for domain restrictions in custom issuer spec
```

```feature user
support for providing private key for custom issuer via shoot resource reference
```

### CODE REVIEW

**Review by mandelsoft — APPROVED** [review]

File: pkg/controller/actuator.go [hunk 1]
```diff
@@ -180,6 +207,22 @@ func (a *actuator) createIssuerValues(issuers ...service.IssuerConfig) ([]map[st
...
 
+func (a *actuator) lookupReferencedSecret(cluster *controller.Cluster, refname string) string {
+	if cluster.Shoot != nil {
+		for _, ref := range cluster.Shoot.Spec.Resources {
+			if ref.Name == refname {
+				if ref.ResourceRef.Kind != "Secret" {
+					a.logger.Info("invalid referenced resource, expected kind Secret, not %s: %s", ref.ResourceRef.Kind, refname)
+					return "invalid-kind"
+				}
+				return "ref-" + ref.ResourceRef.Name
```
Comment (timuthy) [inline pkg/controller/actuator.go:218]:
```suggestion
				return "v1beta1constants.ReferencedResourcesPrefix" + ref.ResourceRef.Name
```

**Review by timuthy — CHANGES_REQUESTED** [review]
Thank you, looks good in general. I wonder if we should now wait until all issuers get ready? I know we haven't done it yet but as we accept more user input the readiness of an issuer can really suffer and there will barely be no feedback since issuers are not reflected in the shoot cluster. WDYT?

File: example/30-extension.yaml [hunk 2]
```diff
@@ -13,7 +13,21 @@ spec:
...
+      externalAccountBinding:
+        keyID: aaaabbbbccccddddeee
+        keySecretName: issuer-custom-eab-hmackey # referenced resource, the "MAC key" must be stored in the secret at `data.hmacKey`
+      skipDNSChallengeValidation: true
+      domains:
+        include:
+          - sub1.my-domain.com
+          - sub2.my-domain.com
+        exclude:
+          - private.sub1.my-domain.com
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply example/30-extension.yaml:29]: done

File: pkg/controller/actuator.go (hunk 1 above)
Comment (MartinWeindel) [reply pkg/controller/actuator.go]: fixed

**Comment by MartinWeindel** [comment]
> I wonder if we should now wait until all issuers get ready? I know we haven't done it yet but as we accept more user input the readiness of an issuer can really suffer and there will barely be no feedback since issuers are not reflected in the shoot cluster. WDYT?

It's a good idea to check the status of all issuers. I wonder if it wouldn't make more sense to have some kind of health check? Anyhow this needs some more testing. I have one stakeholder who is waiting for the setting of the privateKey for a custom issue via a resource reference. May it is better to open an issue and deal with it in another PR?

**Comment by MartinWeindel** [comment]
@timuthy 
Extending the health check was easier than expected.
The health check for the condition type `ControlPlaneHealthy` now also includes a status check of all issuers in the control plane. It reports issuer state and message if state is != 'Ready'

**Review by timuthy — APPROVED** [review]
/lgtm

Side note: We now have an indirect dependency to github.com/gardener/controller-manager-library which will on the long run cause conflicts with the Controller-Runtime and the K8s versions each lib uses. We should try to get rid of this again, but nothing I see critical for now.

**Pull request merged**